### PR TITLE
Implemented initial event-engine based on common EventProcessor

### PR DESCRIPTION
**Pull request by itzg**
# Resolves

https://jira.rax.io/browse/SALUS-1078

# What

Re-implemented the event-engine to be completely dependency free and self contained by using the `EventProcessor` declared in salus-event-engine-common.

# Depends on

- https://github.com/racker/salus-event-engine-common/pull/25
- https://github.com/racker/salus-telemetry-model/pull/177
- https://github.com/racker/salus-common/pull/83

# TODO

- Needs more unit tests
- Needs integration testing

### CODE REVIEW

File: src/main/java/com/rackspace/salus/eventengine/lombok.config [hunk 1]
```diff
@@ -0,0 +1,2 @@
+config.stopbubbling = true
+lombok.accessors.chain = true
```
Comment (shabd67) [inline src/main/java/com/rackspace/salus/eventengine/lombok.config:2]: are these for chained setters?

File: src/main/java/com/rackspace/salus/eventengine/services/EventTaskLoader.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  @Autowired
+  public EventTaskLoader(EventEngineTaskRepository taskRepository) {
+    this.taskRepository = taskRepository;
+  }
+
+  @Transactional(readOnly = true)
+  public void loadAll(Consumer<EventEngineTask> handler) {
+    final Instant startTime = Instant.now();
+    try (Stream<EventEngineTask> taskStream = taskRepository.streamAll()) {
```
Comment (rohitsngh27) [inline src/main/java/com/rackspace/salus/eventengine/services/EventTaskLoader.java:44]: Does these tasks created from event-management?

File: src/main/java/com/rackspace/salus/eventengine/lombok.config (hunk 1 above)
Comment (itzg) [reply src/main/java/com/rackspace/salus/eventengine/lombok.config:2]: Yes

File: src/main/java/com/rackspace/salus/eventengine/services/EventContextResolver.java [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+  int getContextCount() {
+    return contexts.size();
+  }
+
+  public void registerTask(EventEngineTask task) {
+    log.debug("Registering task={}", task);
+    final TasksKey key = new TasksKey(
+        task.getTenantId(), task.getTaskParameters().getMetricGroup());
+
+    tasks.computeIfAbsent(key, unused -> new CopyOnWriteArraySet<>())
```
[the remark this answers is not in the document]
Comment (itzg) [reply src/main/java/com/rackspace/salus/eventengine/services/EventContextResolver.java:110]: I needed a thread-safe `Set` here and out of the two choices in the [concurrent package](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/package-summary.html), this one seemed best for read-heavy access.

File: src/main/java/com/rackspace/salus/eventengine/services/EventTaskLoader.java (hunk 2 above)
Comment (itzg) [reply src/main/java/com/rackspace/salus/eventengine/services/EventTaskLoader.java:44]: Yes

File: README.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# Salus Event Engine
+
+This application is responsible for consuming metrics, locating matching event tasks, updating the state tracking of located tasks, and producing event notifications.
```
Comment (harageth) [inline README.md:3]: README is pretty light. But I am good with leaving it as is and coming back to it later as we work on this, or even in just a documentation PR.

Comment (itzg) [reply README.md:3]: The power of a PoC ;) But yes, a documentation PR for this README or the docs repo is a good idea for a follow up.

File: src/main/java/com/rackspace/salus/eventengine/services/EventContextResolver.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (harageth) [reply src/main/java/com/rackspace/salus/eventengine/services/EventContextResolver.java:110]: And this `CopyOnWriteArraySet` is going to remain small? Its not clear to me yet how its being used.

File: src/main/java/com/rackspace/salus/eventengine/services/EventTaskLoader.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+  private final EventEngineTaskRepository taskRepository;
+
+  @Autowired
+  public EventTaskLoader(EventEngineTaskRepository taskRepository) {
+    this.taskRepository = taskRepository;
+  }
+
+  @Transactional(readOnly = true)
+  public void loadAll(Consumer<EventEngineTask> handler) {
```
Comment (harageth) [inline src/main/java/com/rackspace/salus/eventengine/services/EventTaskLoader.java:42]: Did I miss a startup call to this? Or are we really only reading tasks from Kafka?

File: src/main/java/com/rackspace/salus/eventengine/services/PartitionFinder.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.kafka.core.ConsumerFactory;
+import org.springframework.stereotype.Component;
+import org.springframework.stereotype.Service;
+
+/**
+ * Allows for locating partitions of requested topics in order to allow kafka consumers
+ * in the same group to process all messages in a broadcast-way.
+ * Refer to
+ * <a href="https://docs.spring.io/spring-kafka/reference/html/#tip-assign-all-parts">this section</a>
```
Comment (harageth) [inline src/main/java/com/rackspace/salus/eventengine/services/PartitionFinder.java:29]: This is just a comment for me to come back and look at this to understand it better.

File: src/main/java/com/rackspace/salus/eventengine/services/EventTaskLoader.java (hunk 5 above)
Comment (itzg) [reply src/main/java/com/rackspace/salus/eventengine/services/EventTaskLoader.java:42]: Yes, you did. `EventTaskListener.init` invoked via `@PostConstruct` calls `EventTaskListener.loadTasks` and calls `EventTaskLoader.loadAll` just before starting kafka message listeners.

File: src/main/java/com/rackspace/salus/eventengine/services/EventContextResolver.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (itzg) [reply src/main/java/com/rackspace/salus/eventengine/services/EventContextResolver.java:110]: I'm guessing the "copy" part sounds scary, but array copies are very efficient memory operations. An alternative would have been wrapping a `HashSet` in `Collections.synchronizedSet`, but then the cost of a synchronization monitor has to paid with every lookup on every metric matching the `TasksKey`.

In any case these sets will tend to be fairly small since the hash key (`TasksKey`) is tenant + metric group, where a metric group basically equates to a monitor type or check type in v1 terms. In most cases users will probably have 1 or 2 tasks that fall into each key.

The `tasks` mapping is used as the first level lookup as metrics come in so that the set of label selectors to be evaluated can be efficiently narrowed down.

[the remark this answers is not in the document]
Comment (harageth) [reply src/main/java/com/rackspace/salus/eventengine/services/EventContextResolver.java:110]: I was asking because the CopyOnWrite documentation states that the Set should remain small.

[the remark this answers is not in the document]
Comment (itzg) [reply src/main/java/com/rackspace/salus/eventengine/services/EventContextResolver.java:110]: Ah, that's a good reason to ask. I'm starting to wonder if I even needed a `Set` here or if just a `List` would be sufficient. I might be pre-optimizing the add/remove operations where instead the traversal is more common. I'll explore some of the other concurrent-safe collections...

File: src/main/java/com/rackspace/salus/eventengine/services/EventContextResolver.java [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    final Set<EventEngineTask> candidateTasks = tasks
+        .get(new TasksKey(metric.getTenantId(), metric.getMetricGroup()));
+
+    if (candidateTasks != null) {
+      candidateTasks.stream()
+          // ...then narrow down by label selectors
+          .filter(task -> matchesLabelSelector(task.getTaskParameters(), metric.getLabels()))
+          .forEach(task -> {
+
+            final List<Entry<String, String>> groupingLabels = buildGroupingLabels(metric, task);
```
Comment (harageth) [inline src/main/java/com/rackspace/salus/eventengine/services/EventContextResolver.java:166]: Ahhhh this must be where you are building the larger lists based on the routing from kafka.

File: src/main/java/com/rackspace/salus/eventengine/services/EventContextResolver.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (itzg) [reply src/main/java/com/rackspace/salus/eventengine/services/EventContextResolver.java:110]: I'm now thinking I'll use [ConcurrentSkipListMap](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ConcurrentSkipListMap.html) since the `unregisterTask` is actually removing via iteration with the `removeIf` call. In `process` I'll make a small change to stream the entrySet.

[the remark this answers is not in the document]
Comment (itzg) [reply src/main/java/com/rackspace/salus/eventengine/services/EventContextResolver.java]: and it's changed

File: src/main/java/com/rackspace/salus/eventengine/services/EventContextResolver.java (hunk 7 above)
Comment (itzg) [reply src/main/java/com/rackspace/salus/eventengine/services/EventContextResolver.java:174]: Correct. So, even if the user wants a broadly-selecting task "notify me when the cpu usage of any of my linux servers goes over 80%", then a grouping label list of "resource_id" will indirectly line up with [the kafka message key and partitioning being based on tenantId, resourceId, measurementName](https://github.com/racker/salus-telemetry-ambassador/blob/16906c8eac1adb94dd99238bf0fa00080379db85/src/main/java/com/rackspace/salus/telemetry/ambassador/services/MetricRouter.java#L184).

In other words, it's the same concept as how we used to use [kapacitor's groupBy node](https://docs.influxdata.com/kapacitor/v1.5/nodes/group_by_node/).

**Review by harageth — APPROVED** [review]

**Comment by harageth** [comment]
So I gave approval but we should track the rest of the work that needs to be done here. 
Since this is a POC the important thing for us to test is going to be load testing. As such I am fine with the limited unit tests for this PR. 

As geoff pointed out we need to do some integration testing (manually?) to make sure this is working end to end. And once that is done we should start load testing immediately. 

If that goes well we need to do some documentation and unit testing work here and then after those we can look at other features that we may need. I want the documentation and unit testing done first so that we don't get behind and create a huge amount of work in the future.

**Review by harageth — DISMISSED** [review]

**Comment by itzg** [comment]
With the latest changes I have been able to complete integration testing on my local system by observing these types of notification messages in the `salus.event-notifications.json` topic:

```json
{"id":"tenant-00:test:resource-00-01:23a186e7-39b5-4bde-bd02-4630570d676b:1611171300:WARNING","tenantId":"tenant-00","timestamp":1611171522.713684000,"taskId":"23a186e7-39b5-4bde-bd02-4630570d676b","state":"WARNING","previousState":"OK","observations":[{"zone":"_LOCAL_","state":"WARNING"}],"message":"usage_user of 22 is above warning levels","metricGroup":"cpu","metrics":{"usage_system":179,"usage_user":22,"usage_idle":7},"groupingLabels":[{"system_env":"test"},{"resource_id":"resource-00-01"}],"labels":{"system_arch":"x64","system_os":"windows","resource_id":"resource-00-01","system_env":"test"}}
{"id":"tenant-00:test:resource-00-01:23a186e7-39b5-4bde-bd02-4630570d676b:1611171300:OK","tenantId":"tenant-00","timestamp":1611171532.714328000,"taskId":"23a186e7-39b5-4bde-bd02-4630570d676b","state":"OK","previousState":"WARNING","observations":[{"zone":"_LOCAL_","state":"OK"}],"message":null,"metricGroup":"cpu","metrics":{"usage_system":170,"usage_user":5,"usage_idle":20},"groupingLabels":[{"system_env":"test"},{"resource_id":"resource-00-01"}],"labels":{"system_arch":"x64","system_os":"windows","resource_id":"resource-00-01","system_env":"test"}}
```

where the task created using:
```json
{
	"name": "cpu from MetricsGen",
  "monitoringSystem": "salus",
	"taskParameters": {
    "metricGroup": "cpu",
		"labelSelector": {
			"system_env": "test"
		},
    "groupBy": ["resource_id"],
		"stateExpressions": [
			{
				"state": "CRITICAL",
				"message": "usage_user of #{usage_user} is critical",
				"expression": {
					"type": "comparison",
					"comparator": ">",
					"input": "usage_user",
					"comparisonValue": 50
				}
			},
      {
				"state": "WARNING",
				"message": "usage_user of #{usage_user} is above warning levels",
				"expression": {
					"type": "comparison",
					"comparator": ">",
					"input": "usage_user",
					"comparisonValue": 20
				}
			}
		]
	}
}
```

**Review by harageth — APPROVED** [review]

**Review by rohitsngh27 — APPROVED** [review]
